include/bout/mesh.hxx

```diff
@@ -756,5 +800,201 @@ private:
+  ASSERT1(vel.isAllocated());
+
+  // Check the input data is valid
+  {
+    TRACE("Checking inputs");
+    checkData(f);
+    checkData(vel);
+  }
+
+  // Define properties of this approach
+  const CELL_LOC allowedStaggerLoc = getAllowedStaggerLoc<direction>();
+
+  // Handle the staggering
+  const CELL_LOC inloc = f.getLocation(); // Input locations
+  const CELL_LOC vloc = vel.getLocation();
+  if (outloc == CELL_DEFAULT)
+    outloc = inloc;
+  const STAGGER stagger = getStagger(vloc, inloc, outloc, allowedStaggerLoc);
+
+  // Check for early exit
+  const int nPoint = getNpoints<direction>();
 
+  if (nPoint == 1) {
+    auto tmp = T(0., f.getMesh());
+    tmp.setLocation(outloc);
+    return tmp;
+  }
+
+  // Lookup the method
+  auto& derivativeStore = DerivativeStore<T>::getInstance();
+  typename DerivativeStore<T>::upwindFunc derivativeMethod;
+  if (derivType == DERIV::Upwind) {
+    derivativeMethod = derivativeStore.getUpwindDerivative(method, direction, stagger);
+  } else if (derivType == DERIV::Flux) {
+    derivativeMethod = derivativeStore.getFluxDerivative(method, direction, stagger);
+  } else {
+    throw BoutException("Invalid derivative type in call to indexFlowDerivative.");
+  }
+
+  // Create the result field
+  T result(f.getMesh());
+  result.allocate(); // Make sure data allocated
+  result.setLocation(outloc);
+
+  // Apply method
+  derivativeMethod(vel, f, result, region);
+
+  // Check the result is valid
+  {
+    TRACE("Checking result");
+    checkData(result);
+  }
+
+  return result;
+}
+
+template <typename T, DIRECTION direction, int order>
+T Mesh::indexStandardDerivative(const T& f, CELL_LOC outloc, const std::string& method,
+                                REGION region) const {
+  AUTO_TRACE();
+
+  // Checks
+  static_assert(std::is_base_of<Field2D, T>::value || std::is_base_of<Field3D, T>::value,
+                "indexDDX only works on Field2D or Field3D input");
+  // Check that the mesh is correct
+  ASSERT1(this == f.getMesh());
+  // Check that the input variable has data
+  ASSERT1(f.isAllocated());
+
+  // Check the input data is valid
+  {
+    TRACE("Checking input");
+    checkData(f);
+  }
+
+  // Define properties of this approach
+  const CELL_LOC allowedStaggerLoc = getAllowedStaggerLoc<direction>();
+
+  // Handle the staggering
+  const CELL_LOC inloc = f.getLocation(); // Input location
+  if (outloc == CELL_DEFAULT)
+    outloc = inloc;
+  const STAGGER stagger = getStagger(inloc, outloc, allowedStaggerLoc);
+
+  // Check for early exit
+  const int nPoint = getNpoints<direction>();
+
+  if (nPoint == 1) {
+    auto tmp = T(0., f.getMesh());
+    tmp.setLocation(outloc);
+    return tmp;
+  }
+
+  // Lookup the method
+  auto& derivativeStore = DerivativeStore<T>::getInstance();
+  typename DerivativeStore<T>::standardFunc derivativeMethod;
+
+  if (order == 1) {
+    derivativeMethod = derivativeStore.getStandardDerivative(method, direction, stagger);
```

Comment:
As it is indeed known at compile time there should be no branching cost, just developer cost of reading these lines! Moving order into the store stuff would break some of the symmetry with respect to the other derivative types I think.